pdr_backend/lake/partitioned_data_store.py

```diff
@@ -0,0 +1,114 @@
+        """
+
+        view_name = self._generate_view_name(self.base_directory + dataset_identifier)
+        # Register the DataFrame as an Arrow table temporarily
+        # arrow_table = df.to_arrow()
+        self.duckdb_conn.register(view_name, df)
+
+    @enforce_types
+    def finalize_data(self, dataset_identifier: str):
+        """
+        Finalizes the data by writing the in-memory dataset to partitioned Parquet files.
+        @arguments:
+            dataset_identifier - A unique identifier for the dataset.
+        """
+
+        view_name = self._generate_view_name(self.base_directory + dataset_identifier)
+        query = f"SELECT * FROM {view_name}"
+        result_df = self.duckdb_conn.execute(query).df()
+
+        # Convert to Polars DataFrame for partitioned write
+        df = pl.DataFrame(result_df)
+
+        lazy_df = df.lazy()
+        # Determine partitions based on a timestamp column, assuming 'timestamp' column exists
+        lazy_df = lazy_df.with_columns(
+            [pl.from_epoch("timestamp", time_unit="s").alias("date")]
+        )
+        df = lazy_df.collect()
+
+        partitions = (
+            lazy_df.select(
+                pl.col("date").dt.year().alias("year"),
+                pl.col("date").dt.month().alias("month"),
+                pl.col("date").dt.day().alias("day"),
+            )
+            .unique()
+            .collect()
+        )
+
+        for partition in partitions.rows():
+            year, month, day = partition
+            partition_path = os.path.join(
+                self.base_directory,
+                dataset_identifier,
+                f"year={year}",
+                f"month={month}",
+            )
+            os.makedirs(partition_path, exist_ok=True)
+            partition_df = df.filter(
+                (pl.col("date").dt.year() == year)
+                & (pl.col("date").dt.month() == month)
+            )
+            file_path = os.path.join(partition_path, f"{day}.parquet")
+            partition_df.write_parquet(file_path)
+
+    @enforce_types
+    def query_data(self, dataset_identifier: str, query: str) -> pl.DataFrame:
+        """
+        Execute a SQL query across partitioned Parquet files using DuckDB.
+        @arguments:
+            dataset_identifier - A unique identifier for the dataset.
+            query - The SQL query to execute. The query must contain a {view_name} placeholder.
+
+        @returns:
+            pl.DataFrame - The result of the query.
+        """
+
+        dataset_path = os.path.join(self.base_directory, dataset_identifier)
+        view_name = self._generate_view_name(dataset_path)
+
+        # DuckDB automatically recognizes the partition scheme when reading from a directory
+        self.duckdb_conn.execute(
+            f"CREATE OR REPLACE VIEW {view_name} AS SELECT * FROM read_parquet('{dataset_path}/*/*/*.parquet', hive_partitioning=1)"  # pylint: disable=line-too-long
```

Comment:
Hi, we do not want to be loading all the data into memory. This will eventually become too big.

Please do the following:
(1) update query_data() to run queries against the self.duckdb persistent tables.
(2) create a different query function to read from polars (look for Norbert's Table update) such that we (a) generate the `parquet-hive-partition.parquet` path and then (b) return a pl.LazyFrames(path) such that the data is not loaded onto memory and a polars lazy interface is returned
(3) update current query function to read from duck db peristent tables

In the end:
1. saving to parquet will be done in 2 steps: (a) via the db by registering/appending, then (b) by calling finalize()
2. most of the ETL work will be done via sql, on the db, such that only the required records are reprocessed
3. you can then query clean, final tables, or just the latest records w/ sql, for immediate results
4. as a final step of the ETL, we dump the latest db/records to parquet